core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -186,8 +186,12 @@ class Partition(val topic: String,
       // record the epoch of the controller that made the leadership decision. This is useful while updating the isr
       // to maintain the decision maker controller's epoch in the zookeeper path
       controllerEpoch = partitionStateInfo.controllerEpoch
-      // add replicas that are new
-      allReplicas.foreach(replica => getOrCreateReplica(replica))
+      // add replicas that are new and reset lastCatchUpTime of non-isr replicas
+      allReplicas.foreach(replica => {
+        val r = getOrCreateReplica(replica)
+        if (!partitionStateInfo.isr.contains(replica))
+          r.resetLastCatchUpTime()
+      })
```

Comment:
Yes, it makes sense to lastFetchTimeMs to curTime. Could we just leave lastFetchLeaderLogEndOffset to 0 or -1?

Another thing is that we probably only need to do this optimization when the leader is indeed changing.